### PR TITLE
!!! BUGFIX: DocumentUriProjection doesn't respect fallback nodes; NodePropertiesWereSet contains affectedDimensionSpacePoints

### LINKED ISSUE #4265: 9.0: Add affected dimension space points to NodePropertiesWereSet event types
Other than the related events the events NodePropertiesWereSet and NodeReferencesWereSet don't contain the dimension space points that were affected.

This is required to fix https://github.com/neos/neos-development-collection/issues/4256

As a result it is not possible to tell which node variants were updated without tracking the dimension fallback chain

Original issue: https://github.com/neos/contentrepository-development-collection/issues/163

### LINKED ISSUE #4270: 9.0 BUG: Shortcut settings are not updated in `DocumentUriPathProjection` table on document creation
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

If a user creates a new document of type shortcut the changes on `targetMode` wont be persisted in `cr_default_p_neos_documenturipath_uri`. It will always contain `parentNode` on first creation although the data in `cr_default_p_graph_node` is updated correctly.

As I debugged the data in `cr_default_p_neos_documenturipath_uri` I did disable the cache in `DocumentUriPathFinder`. In this case all new shortcut nodes were updated in the `DocumentUriPathProjection`.

But this behaviour couldn't be reproduced with `targetMode` set to `firstChildNode`.

Example from the database:
| uripath | shortcuttarget |
| - | - |
| test-shortcut-cached | {"mode":"parentNode","target":"https:\/\/www.google.com"} | 
| test-shortcut-firstChild-cached | {"mode":"firstChildNode","target":null} |
| subfolder/test-shortcut-cached | {"mode":"parentNode","target":"node:\/\/28255929-524f-4461-b125-78a11e1559a4"} | 
| test-shortcut-uncached | {"mode":"selectedTarget","target":"https:\/\/www.google.com"} | 

### Expected Behavior

Switching `targetMode` on document node creation to `selectedTarget` should be persisted in projection table.

### Steps To Reproduce

1. Create a node of type shortcut
2. Change `targetMode` to `selectedTarget`
3. Select a document node for `target` or choose an external link
4. Publish

### Environment

```markdown
- Flow: 8.3
- Neos: 9.0
- PHP: 8.1
```

### Anything else?

⚠️ Testing this with: #4266 

Relates: #4253

### PR DESCRIPTION
**Pull request by skurfuerst**
Extending the NodePropertiesWereSet event is useful for some projections which are not interested in OriginDimensionSpacePoints, but need to know where a property value is applied (like the DocumentUriPathProjection).

This change is BREAKING because it updates the event payload of NodePropertiesWereSet. To update your event store, run:

./flow migrateEvents:fillAffectedDimensionSpacePointsInNodePropertiesWereSet

We also fix the DocumentUriPathProjection alongside.

Resolves: #4265
Resolves: #4256
Resolves: #4270 
Resolves: #4253

### CODE REVIEW

**Comment by dlubitz** [comment]
Looks good by reading. Will try it within next days. 
Just one test is failing due to the changes.

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Event/NodePropertiesWereSet.php [hunk 1]
```diff
@@ -41,7 +42,9 @@ final class NodePropertiesWereSet implements
     public function __construct(
         public readonly ContentStreamId $contentStreamId,
         public readonly NodeAggregateId $nodeAggregateId,
-        public readonly OriginDimensionSpacePoint $originDimensionSpacePoint, // TODO
+        public readonly OriginDimensionSpacePoint $originDimensionSpacePoint,
+        /** the covered dimension space points for this modification - i.e. where this change is visible */
+        public readonly DimensionSpacePointSet $affectedDimensionSpacePoints,
```
Comment (ahaeslich) [inline Neos.ContentRepository.Core/Classes/Feature/NodeModification/Event/NodePropertiesWereSet.php]: There are two other places in the current code base where `NodePropertiesWereSet` is initialized, which are not addressed in this MR:

- `Neos\ContentRepository\StructureAdjustment\Adjustment\PropertyAdjustment->publishNodePropertiesWereSet()`
- `Neos\ContentRepository\LegacyNodeMigration\NodeDataToEventsProcessor->createNodeVariant()`

File: Neos.Neos/Tests/Behavior/Features/Bootstrap/FeatureContext.php [hunk 2]
```diff
@@ -111,11 +120,80 @@ public function __construct()
         $this->objectManager = self::$bootstrap->getObjectManager();
         $this->environment = $this->objectManager->get(Environment::class);
 
+        $this->nodeAuthorizationService = $this->objectManager->get(AuthorizationService::class);
```
Comment (ahaeslich) [inline Neos.Neos/Tests/Behavior/Features/Bootstrap/FeatureContext.php]: This is a duplicate of the next line
```suggestion
```

**Review by ahaeslich — CHANGES_REQUESTED** [review]

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Event/NodePropertiesWereSet.php (hunk 1 above)
Comment (skurfuerst) [reply Neos.ContentRepository.Core/Classes/Feature/NodeModification/Event/NodePropertiesWereSet.php]: good catch :) thanks!

Comment (skurfuerst) [reply Neos.ContentRepository.Core/Classes/Feature/NodeModification/Event/NodePropertiesWereSet.php]: done

File: Neos.Neos/Tests/Behavior/Features/Bootstrap/FeatureContext.php (hunk 2 above)
Comment (skurfuerst) [reply Neos.Neos/Tests/Behavior/Features/Bootstrap/FeatureContext.php]: done

**Comment by skurfuerst** [comment]
let's hope CI runs through now, then this would be merge ready IMHO. Tests locally run though :)

Thanks everybody!

**Comment by ahaeslich** [comment]
> let's hope CI runs through now

It won't due to the recent dependency changes in https://github.com/neos/flow-development-collection/commits/9.0

**Review by ahaeslich — APPROVED** [review]
Tested the changes in our project 👍 💖

**Pull request merged**